src/libshared_ptr/arc.rs

```diff
@@ -528,6 +540,185 @@ impl<T: ?Sized> Arc<T> {
     }
 }
 
+impl<T> Arc<[T]> {
+    // Returns size and alignment of `ArcInner<[T]>` with `len` elements
+    fn size_align_of(len: usize) -> (usize, usize) {
+        unsafe {
+            let ptr: *mut ArcInner<[T]> =
+                mem::transmute([mem::align_of::<ArcInner<[T; 1]>>(), len]);
+
+            // FIXME(custom-DST): creating this invalid &[T] is dubiously defined,
+            // we should have a better way of getting the size/align
+            // of a DST from its unsized part.
+            (size_of_val(&*ptr), align_of_val(&*ptr))
+        }
+    }
+
+    unsafe fn allocate_slice(len: usize) -> *mut ArcInner<[T]> {
+        let (size, align) = Self::size_align_of(len);
+
+        let ptr = allocate(size, align);
+
+        if ptr.is_null() {
+            oom();
+        }
+
+        Self::init_arc_inner(ptr, len)
+    }
+
+    unsafe fn init_arc_inner(ptr: *mut u8, len: usize) -> *mut ArcInner<[T]> {
+        let ptr: *mut ArcInner<[T]> = mem::transmute([ptr as usize, len]);
+
+        ptr::write(&mut (*ptr).strong, atomic::AtomicUsize::new(1));
+        ptr::write(&mut (*ptr).weak, atomic::AtomicUsize::new(1));
+
+        ptr
+    }
+
+    fn from_vec(mut v: Vec<T>) -> Arc<[T]> {
+        unsafe {
+            let ptr = Self::allocate_slice(v.len());
+
+            ptr::copy_nonoverlapping(
+                v.as_ptr(),
+                &mut (*ptr).data as *mut [T] as *mut T,
+                v.len());
+
+            // Allow the Vec to free its memory, but not destroy its contents
+            v.set_len(0);
+
+            Arc { ptr: Shared::new(ptr) }
+        }
+    }
+}
+
+// Returns the offset of a field with `align` alignment,
+// placed after at least `start` bytes
+fn offset_of_aligned(start: usize, align: usize) -> usize {
+    // Align for zero-sized values is 1,
+    // so we shouldn't need to check whether align == 0
+    if start % align == 0 {
+        start
+    } else {
+        start - (start % align) + align
+    }
+}
+
+impl<T: ?Sized> Arc<T> {
+    fn from_box(v: Box<T>) -> Arc<T> {
```

Comment:
@murarth 

The `align_of` trick is a bit ugly, but I think it's better than hardcoding layout calculations (as `allocate_for_ptr` does) and is probably going to be well-defined.

I think the "official" way of creating "franken-raw-pointers" is by transmuting to `std::raw::TraitObject` rather than odd `ptr::write`s etc.